refaster-support/src/main/java/tech/picnic/errorprone/refaster/AnnotatedCompositeCodeTransformer.java

```diff
@@ -0,0 +1,154 @@
+package tech.picnic.errorprone.refaster;
+
+import static com.google.common.base.Preconditions.checkState;
+import static com.google.errorprone.BugPattern.SeverityLevel.ERROR;
+import static com.google.errorprone.BugPattern.SeverityLevel.SUGGESTION;
+import static com.google.errorprone.BugPattern.SeverityLevel.WARNING;
+import static tech.picnic.errorprone.refaster.annotation.OnlineDocumentation.NESTED_CLASS_URL_PLACEHOLDER;
+import static tech.picnic.errorprone.refaster.annotation.OnlineDocumentation.TOP_LEVEL_CLASS_URL_PLACEHOLDER;
+
+import com.google.auto.value.AutoValue;
+import com.google.common.base.Splitter;
+import com.google.common.collect.Comparators;
+import com.google.common.collect.ImmutableClassToInstanceMap;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.Iterators;
+import com.google.errorprone.BugPattern.SeverityLevel;
+import com.google.errorprone.CodeTransformer;
+import com.google.errorprone.CompositeCodeTransformer;
+import com.google.errorprone.DescriptionListener;
+import com.google.errorprone.ErrorProneOptions;
+import com.google.errorprone.matchers.Description;
+import com.sun.source.util.TreePath;
+import com.sun.tools.javac.util.Context;
+import java.io.Serializable;
+import java.lang.annotation.Annotation;
+import java.util.Iterator;
+import java.util.Optional;
+import java.util.function.Function;
+import tech.picnic.errorprone.refaster.annotation.OnlineDocumentation;
+import tech.picnic.errorprone.refaster.annotation.Severity;
+
+/**
+ * A {@link CompositeCodeTransformer} that can be augmented with annotations.
+ *
+ * <p>The content of the annotations can be used for enriched compilation warnings or errors, and
+ * documentation purposes.
```

Comment:
Thanks for the Javadocs @rickie! I jotted something down in a private note when I had a sudden brainpulse a while ago:

> A composite code transformer that enhances the description of matches from Refaster rules based on optionally available annotations.

Placing it here to see if (part of it) might be nice to still adopt. Specifically, ^ is more focused around _current_ functionality and Refaster rules specifically. Feel free to consider `s/enhances/augments/` and `s/rules/templates/`.

Reply:
While debugging, I got a better understanding of how the annotations are retrieved and passed down. As such, I understood that the annotations available inside _this_ `AnnotatedCompositeCodeTransformer` are (per annotation type) of the first available parent class (i.e. of one of the encompassing parent template collections) of the `RefasterRule` `CodeTransformer`s herein. Here, we augment descriptions on matches per respective `CodeTransformer` based on annotations available on said `CodeTransformer` and otherwise use the encompassing annotations mentioned above (not that this class knows, it just has been given a list of fallback annotations to rely upon).

As such, I'm wondering whether we should clarify this Javadoc accordingly as 'the annotations' is currently a bit ambigious over which annotations in the context of (`Composite`) `CodeTransformer`s we are talking about. Committed a suggestion, PTAL. Not entirely happy with 'as a fallback' but was less happy with words such as 'otherwise' or 'default'.

Also not sure if we should mention the annotations in the summary given the class' name, but left it for now. I did change the summary line as semantically it was saying that the annotations could augment this `CompositeCodeTransformer`. Also removed the XXX for lack of a better alternative on my side too.

Feel free to revert or change if I did not get the design of this class correctly :sweat_smile: